pcsx2/DEV9/InternalServers/DNS_Server.cpp

```diff
@@ -0,0 +1,435 @@
+		//Different elements in the same container can be modified concurrently by different threads
+		for (size_t i = 0; i < dnsQuestions.size(); i++)
+			answers[dnsQuestions[i]] = {0};
+	}
+
+	int DNS_Server::DNS_State::AddAnswer(const std::string& answer, IP_Address address)
+	{
+		answers[answer] = address;
+		return --counter;
+	}
+	int DNS_Server::DNS_State::AddNoAnswer(const std::string& answer)
+	{
+		return --counter;
+	}
+
+	std::unordered_map<std::string, IP_Address> DNS_Server::DNS_State::GetAnswers()
+	{
+		return answers;
+	}
+
+	DNS_Server::DNS_Server(std::function<void()> receivedcallback)
+		: callback{receivedcallback}
+	{
+#ifdef _WIN32
+		/* Use the MAKEWORD(lowbyte, highbyte) macro declared in Windef.h */
+		WORD wVersionRequested = MAKEWORD(2, 2);
+
+		WSADATA wsaData{0};
+		const int err = WSAStartup(wVersionRequested, &wsaData);
+		if (err != 0)
+			Console.Error("DEV9: WSAStartup failed with error: %d\n", err);
+		else
+			wsa_init = true;
+#endif
+	}
+
+	//We remap 127.0.0.1 to the PC's IP address
+	//We specificly use the address assigned to
+	//the adapter we are using
+#ifdef _WIN32
+	void DNS_Server::Init(PIP_ADAPTER_ADDRESSES adapter)
+#elif defined(__POSIX__)
+	void DNS_Server::Init(ifaddrs* adapter)
+#endif
+	{
+		localhostIP = {127, 0, 0, 1};
+		//Find IPv4 Address
+#ifdef _WIN32
+		PIP_ADAPTER_UNICAST_ADDRESS address = nullptr;
+		if (adapter != nullptr)
+		{
+			PIP_ADAPTER_ADDRESSES info = (PIP_ADAPTER_ADDRESSES)adapter;
+			address = info->FirstUnicastAddress;
+			while (address != nullptr && address->Address.lpSockaddr->sa_family != AF_INET)
+				address = address->Next;
+		}
+
+		if (address != nullptr)
+		{
+			sockaddr_in* sockaddr = (sockaddr_in*)address->Address.lpSockaddr;
+			localhostIP = *(IP_Address*)&sockaddr->sin_addr;
+		}
+		else
+			Console.Error("DEV9: Failed To Get Adapter IP");
+#elif defined(__POSIX__)
+		sockaddr* address = nullptr;
+		if (adapter != nullptr)
+		{
+			ifaddrs* info = (ifaddrs*)adapter;
+			if (info->ifa_addr != nullptr && info->ifa_addr->sa_family == AF_INET)
+				address = info->ifa_addr;
+		}
+
+		if (address != nullptr)
+		{
+			sockaddr_in* sockaddr = (sockaddr_in*)address;
+			localhostIP = *(IP_Address*)&sockaddr->sin_addr;
+		}
+		else
+			Console.Error("DEV9: Failed To Get Adapter IP");
+#endif
+
+		LoadHostList();
+	}
+
+	void DNS_Server::LoadHostList()
+	{
+		hosts.clear();
+		for (size_t i = 0; i < config.EthHosts.size(); i++)
+		{
+			ConfigHost entry = config.EthHosts[i];
+			if (entry.Enabled)
+				hosts.insert_or_assign(entry.Url, *(IP_Address*)entry.Address);
+		}
+	}
+
+	UDP_Packet* DNS_Server::Recv()
+	{
+		UDP_Packet* retPay;
+		if (dnsQueue.Dequeue(&retPay))
+		{
+			outstandingQueries--;
+			return retPay;
+		}
+		return nullptr;
+	}
+
+	bool DNS_Server::Send(UDP_Packet* payload)
+	{
+		PayloadPtr* udpPayload = static_cast<PayloadPtr*>(payload->GetPayload());
+		DNS_Packet* dns = new DNS_Packet(udpPayload->data, udpPayload->GetLength());
```

Comment:
Would this and other allocations benefit from std::shared_ptr or std::unique_ptr to ensure that the memory is properly managed across all paths?

Reply:
This particular line didn't need to be an allocation and has been corrected

At a glance I'm not sure what would most benefit from using smart pointers.

Reply:
Re-checking I don't see anything obvious beyond the line I commented on. You may resolve this.